parser/parser.go

```diff
@@ -1,15 +1,215 @@
 package parser
 
 import (
+	"fmt"
+	"strconv"
+
 	"github.com/NeowayLabs/abad/ast"
 	"github.com/NeowayLabs/abad/internal/utf16"
+	"github.com/NeowayLabs/abad/lexer"
+	"github.com/NeowayLabs/abad/token"
 )
 
+type (
+	Parser struct {
+		tokens    <-chan lexer.Tokval
+		lookahead []lexer.Tokval
+
+		filename string
+	}
+
+	parserfn func(*Parser) (ast.Node, error)
+)
+
+// used when the tokens is over
+var tokEOF = lexer.Tokval{
```

Comment:
lexer could have a lexer.EOFToken() that returns this EOF token for you =D

Reply:
Yes, I made this in nash but then disliked. It's not a real token, the lexer could inform the end of stream by closing the channel. But maybe makes sense, I don't know.